atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/compact/BackgroundCompactor.java

```diff
@@ -49,40 +50,57 @@
     private final TransactionManager transactionManager;
     private final KeyValueService keyValueService;
     private final LockService lockService;
-    private final Supplier<CompactorConfig> config;
+    private final Supplier<Boolean> inSafeHours;
 
     private Thread daemon;
 
-    public static void createAndRun(TransactionManager transactionManager,
+    public static Optional<BackgroundCompactor> createAndRun(TransactionManager transactionManager,
             KeyValueService keyValueService,
             LockService lockService,
-            Supplier<CompactorConfig> compactorConfigSupplier) {
+            Supplier<Boolean> inSafeHours) {
+        if (!keyValueService.shouldManuallyCompact()) {
+            return Optional.empty();
+        }
+
         BackgroundCompactor backgroundCompactor = new BackgroundCompactor(transactionManager,
                 keyValueService,
                 lockService,
-                compactorConfigSupplier);
+                inSafeHours);
         backgroundCompactor.runInBackground();
+
+        return Optional.of(backgroundCompactor);
     }
 
-    public BackgroundCompactor(TransactionManager transactionManager,
+    private BackgroundCompactor(TransactionManager transactionManager,
             KeyValueService keyValueService,
             LockService lockService,
-            Supplier<CompactorConfig> config) {
+            Supplier<Boolean> inSafeHours) {
         this.transactionManager = transactionManager;
         this.keyValueService = keyValueService;
         this.lockService = lockService;
-        this.config = config;
+        this.inSafeHours = inSafeHours;
     }
 
-    public synchronized void runInBackground() {
+    @Override
+    public void close() {
+        log.info("Closing BackgroundCompactor");
+        daemon.interrupt();
+        try {
+            daemon.join();
+            daemon = null;
```

Comment:
A bit uncertain why we need this assignment. Are we retaining references to the background compactor outside of a TM? Otherwise it looks like the TM after closing should itself be collected, and if that covers all references to this BackgroundCompactor then it should already be garbage-collected.

Reply:
I think a lot of this code was copied from the background sweeper, so may have inherited some foibles.

Reply:
Yep, this code was copied from `BackgroundSweeperImpl`, and previously was copied from $LARGE_INTERNAL_PROJECT back in Nov 2015. Potential tiny task in a few weeks to clean up some of the duplicated code/logic, but not worth it now.